File: src/LCSolution/LC739.java
package LCSolution;

import java.util.HashMap;
import java.util.Map;

public class LC739 {

    class Solution {
        public int[] dailyTemperatures(int[] T) {
            int n = T.length;
            int[] res = new int[n];

            for (int i=n-2;i>=0;i--){

                int addNum = 1;
                int index = i;
                while (addNum!=0){
                    index +=addNum;
                    if (T[index]>T[i]){
                        res[i] = index-i;
                        break;
                    }
                    addNum = res[index];
                }
            }
            return res;
        }
    }
}

File: srcKotlin/LCKotlin/LC739.kt
package LCKotlin

import java.util.Stack

class LC739 {

    fun dailyTemperatures(temperatures: IntArray): IntArray {
        if (temperatures.size == 1) {
            return IntArray(1)
        }
        var res = IntArray(temperatures.size)
        // 单调栈
        var stack: Stack<Int> = Stack()
        for (i in temperatures.indices) {
            while (!stack.empty() && temperatures[i] > temperatures[stack.peek()]) {
                // 当前温度比栈顶温度更高
                var pop = stack.pop()
                res[pop] = i - pop
            }
            stack.push(i)
        }
        return res
    }
}